functions/go/set-labels/transformer/labels.go

```diff
@@ -92,28 +92,162 @@ func (p *LabelTransformer) setLabelsInSpecs(o *fn.KubeObject) error {
 
 // Transform updates the labels in the right path using GVK filter and other configurable fields
 func (p *LabelTransformer) Transform(objects fn.KubeObjects) error {
+	if objects.Len() == 0 || objects[0] == nil {
+		return fmt.Errorf("no resources found")
+	}
 	for _, o := range objects.WhereNot(func(o *fn.KubeObject) bool { return o.IsLocalConfig() }) {
 		// this label need to set for all GVK
-		metaLabelsPath := FieldPath{"metadata", "labels"}
-		updatedLabels, err := updateLabels(&o.SubObject, metaLabelsPath, p.NewLabels, true)
-		if err != nil {
+		if err := p.setMetadataForAll(o); err != nil {
 			return err
 		}
-		p.LogResult(o, metaLabelsPath, updatedLabels)
 		// set other common labels according to specific GVK
-		err = p.setLabelsInSpecs(o)
-		if err != nil {
+		if err := p.setCommonSpecLabels(o); err != nil {
+			return err
+		}
+		// handle special cases when slices are involved
+		if err := p.setLabelsInSlice(o); err != nil {
 			return err
 		}
-		// handle special case with slice
-		if o.IsGVK("apps", "", "StatefulSet") {
-			if o.GetMap("spec") != nil {
-				for _, vctObj := range o.GetMap("spec").GetSlice("volumeClaimTemplates") {
-					updatedLabels, err = updateLabels(vctObj, metaLabelsPath, p.NewLabels, true)
+	}
+	return nil
+}
+
+func (p *LabelTransformer) setMetadataForAll(o *fn.KubeObject) error {
+	metaLabelsPath := FieldPath{"metadata", "labels"}
+	updatedLabels, err := updateLabels(&o.SubObject, metaLabelsPath, p.NewLabels, true)
+	if err != nil {
+		return err
+	}
+	p.LogResult(o, metaLabelsPath, updatedLabels)
+	return nil
+}
+
+// setLabelsInSlice handles the resources that contain slice type
+func (p *LabelTransformer) setLabelsInSlice(o *fn.KubeObject) error {
+	// handle resources that have podSpec struct
+	if err := p.podSpecCheckAndUpdate(o); err != nil {
+		return err
+	}
+	// handle other special case resources
+	if err := p.specialCasesCheckAndUpdate(o); err != nil {
+		return err
+	}
+	return nil
+}
+
+// podSpecCheckAndUpdate updates labels path inside podSpec struct
+func (p *LabelTransformer) podSpecCheckAndUpdate(o *fn.KubeObject) error {
+	if o.IsGVK("", "v1", "ReplicationController") ||
+		o.IsGVK("", "", "Deployment") ||
+		o.IsGVK("", "", "ReplicaSet") ||
+		o.IsGVK("", "", "DaemonSet") ||
+		o.IsGVK("apps", "", "StatefulSet") ||
+		o.IsGVK("batch", "", "Job") {
+		_, exist, _ := o.NestedString(FieldPath{"spec", "template", "spec"}...)
+		if exist {
+			podSpecObj := o.GetMap("spec").GetMap("template").GetMap("spec")
+			if err := p.processPodSpec(podSpecObj, o); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// processPodSpec takes in podSpec object and parse its path, parent kubeObject is also passed in for logging
+func (p *LabelTransformer) processPodSpec(o *fn.SubObject, parentO *fn.KubeObject) error {
```

Comment:
I tried this function in`set-default-labels` I still think it is too tedious to log every resource changes. Can we remove the resource level `LogResult`?

Reply:
By removing the resource level, do you mean only want a general count of how many resources have been updated in total?

Reply:
yeah. `kpt fn` truncates the log if the output is too large. This is bad for PASS function because 1. normally the last log message contains the summary 2. truncated log is more used for FAIL function.  It seems that the set-labels can easily exceed the limitations.